lib/at_host/at_host.c

```diff
@@ -113,75 +122,75 @@ static void cmd_send(struct k_work *work)
 static void uart_rx_handler(u8_t character)
 {
 	static bool inside_quotes;
-	static size_t cmd_len;
-	size_t pos;
+	static size_t at_cmd_len;
 
-	cmd_len += 1;
-	pos = cmd_len - 1;
-
-	/* Handle special characters. */
+	/* Handle control characters */
 	switch (character) {
 	case 0x08: /* Backspace. */
 		/* Fall through. */
 	case 0x7F: /* DEL character */
-		pos = pos ? pos - 1 : 0;
-		at_buf[pos] = 0;
-		cmd_len = cmd_len <= 1 ? 0 : cmd_len - 2;
-		break;
-	case '"':
-		inside_quotes = !inside_quotes;
-		 /* Fall through. */
-	default:
-		/* Detect AT command buffer overflow or zero length */
-		if (cmd_len > AT_MAX_CMD_LEN) {
-			LOG_ERR("Buffer overflow, dropping '%c'\n", character);
-			cmd_len = AT_MAX_CMD_LEN;
-			return;
-		} else if (cmd_len < 1) {
-			LOG_ERR("Invalid AT command length: %d", cmd_len);
-			cmd_len = 0;
-			return;
+		if (at_cmd_len > 0) {
+			at_cmd_len--;
 		}
+		return;
+	}
 
-		at_buf[pos] = character;
-		break;
+	/* Handle termination sequence, if outside quotes */
+	if (!inside_quotes) {
+		switch (term_mode) {
+		case MODE_NULL_TERM:
+			if (character == '\0') {
+				goto send;
+			}
+			break;
+		case MODE_CR:
+			if (character == '\r') {
+				goto send;
+			}
+			break;
+		case MODE_LF:
+			if (character == '\n') {
+				goto send;
+			}
+			break;
+		case MODE_CR_LF:
+			if ((at_buf[at_cmd_len - 1] == '\r') &&
+			    (character == '\n')) {
+				at_cmd_len--; /* Remove preceding CR */
+				goto send;
+			}
+			break;
+		default:
+			LOG_ERR("Invalid termination mode: %d", term_mode);
+			break;
+		}
 	}
 
-	if (inside_quotes) {
+	/* Detect AT command buffer overflow, leaving space for null */
+	if (at_cmd_len + 1 > CONFIG_AT_HOST_CMD_MAX_LEN - 1) {
+		LOG_ERR("Buffer overflow, dropping '%c'\n", character);
 		return;
 	}
 
-	/* Check if the character marks line termination. */
-	switch (term_mode) {
-	case MODE_NULL_TERM:
-		/* Fall through. */
-	case MODE_CR:
-		if (character == termination[term_mode]) {
-			goto send;
-		}
-		break;
-	case MODE_LF:
-		if ((at_buf[pos - 1]) &&
-			character == termination[term_mode]) {
-			goto send;
-		}
-		break;
-	case MODE_CR_LF:
-		if ((at_buf[pos - 1] == '\r') && (character == '\n')) {
-			goto send;
-		}
-		break;
-	default:
-		LOG_ERR("Invalid termination mode: %d", term_mode);
-		break;
+	/* Write character to AT buffer */
+	at_buf[at_cmd_len] = character;
+	at_cmd_len++;
+
+	/* Handle special written character */
+	if (character == '"') {
+		inside_quotes = !inside_quotes;
 	}
 
 	return;
 send:
-	uart_irq_rx_disable(uart_dev);
-	k_work_submit(&cmd_send_work);
-	at_buf_len = cmd_len;
-	cmd_len = 0;
+	uart_irq_rx_disable(uart_dev); /* Block the UART to protect buffer */
+	at_buf[at_cmd_len] = '\0'; /* Terminate the command string */
+
```

Comment:
Have you considered adding an extra check here, to verify that we do not send an empty line? I've noticed that at_cmd drivers stalls if it transmits a sole newline character (probably the modem does not respond for that in any way, and we're stuck in `recv` call). It's quite common for me to do it when running samples with at_host enabled. Not sure though if we should handle this here or in the driver itself.

Reply:
Added a check. I put it at the end so it could not impact the state reset